projection/src/test/scala/akka/projection/r2dbc/DurableStateEndToEndSpec.scala

```diff
@@ -0,0 +1,251 @@
+          system,
+          R2dbcDurableStateStore.Identifier,
+          entityTypeHint,
+          range.min,
+          range.max)
+      val projection = R2dbcProjection
+        .exactlyOnce(
+          projectionId,
+          Some(settings),
+          sourceProvider = sourceProvider,
+          handler = () => handlers(projectionId))
+      spawn(ProjectionBehavior(projection))
+    }.toVector
+  }
+
+  "A R2DBC projection with changesBySlices source" must {
+
+    "handle latest updated state exactlyOnce" in {
+      val numberOfEntities = 20
+      val numberOfChanges = 10 * numberOfEntities
+      val entityTypeHint = nextEntityTypeHint()
+
+      val entities = (0 until numberOfEntities).map { n =>
+        val persistenceId = PersistenceId(entityTypeHint, s"p$n")
+        spawn(DurableStatePersister(persistenceId), s"p$n")
+      }
+
+      var revisionPerEntity = Map.empty[Int, Int]
+
+      // write some before starting the projections
+      (1 to 50).foreach { n =>
+        val p = n % numberOfEntities
+        val revision = revisionPerEntity.getOrElse(p, 0) + 1
+        revisionPerEntity = revisionPerEntity.updated(p, revision)
+        entities(p) ! DurableStatePersister.Persist(s"s$p-$revision")
+      }
+
+      val projectionName = UUID.randomUUID().toString
+      val handlers = createHandlers(projectionName, nrOfProjections = 4)
+      val projections = startProjections(entityTypeHint, projectionName, nrOfProjections = 4, handlers)
+
+      // give them some time to start before writing more events
+      Thread.sleep(500)
+
+      var n = 51
+      while (n <= numberOfChanges) {
+        val p = n % numberOfEntities
+        val revision = revisionPerEntity.getOrElse(p, 0) + 1
+        revisionPerEntity = revisionPerEntity.updated(p, revision)
+        entities(p) ! DurableStatePersister.Persist(s"s$p-$revision")
+
+        // stop projections
+        if (n == numberOfChanges / 2) {
+          val probe = createTestProbe()
+          projections.foreach { ref =>
+            ref ! ProjectionBehavior.Stop
+            probe.expectTerminated(ref)
+          }
+        }
+
+        // resume projections again
+        if (n == (numberOfChanges / 2) + 20)
+          startProjections(entityTypeHint, projectionName, nrOfProjections = 4, handlers)
+
+        if (n % 10 == 0)
+          Thread.sleep(50)
+        else if (n % 25 == 0)
+          Thread.sleep(1500)
+
+        n += 1
+      }
+
+      handlers.foreach { case (projectionId, handler) =>
+        (0 until numberOfEntities).foreach { p =>
+          val persistenceId = PersistenceId(entityTypeHint, s"p$p")
+          val slice = DurableStateSourceProvider2.sliceForPersistenceId(
+            system,
+            R2dbcDurableStateStore.Identifier,
+            persistenceId.id)
+          withClue(s"projectionId $projectionId, persistenceId $persistenceId, slice $slice: ") {
+            if (handler.sliceRange.contains(slice)) {
+              eventually {
```

Comment:
ok, so eventually we see the highest revision, but the earlier updates may be "folded" into fewer than one per update.

Reply:
exactly